src/platform/OpenThread/GenericThreadBorderRouterDelegate.cpp

```diff
@@ -0,0 +1,193 @@
+/*
+ *
+ *    Copyright (c) 2024 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "GenericThreadBorderRouterDelegate.h"
+
+#include <openthread/border_agent.h>
+#include <openthread/dataset.h>
+#include <openthread/error.h>
+#include <openthread/instance.h>
+#include <openthread/ip6.h>
+#include <openthread/link.h>
+#include <openthread/netdiag.h>
+#include <openthread/thread.h>
+#include <openthread/thread_ftd.h>
+
+#include <app-common/zap-generated/cluster-enums.h>
+#include <app/clusters/thread-border-router-management-server/thread-br-delegate.h>
+#include <inet/IPAddress.h>
+#include <lib/core/CHIPError.h>
+#include <lib/support/CodeUtils.h>
+#include <lib/support/Span.h>
+#include <platform/CHIPDeviceEvent.h>
+#include <platform/CHIPDeviceLayer.h>
+#include <platform/KeyValueStoreManager.h>
+#include <platform/PlatformManager.h>
+#include <platform/ThreadStackManager.h>
+
+namespace chip {
+namespace app {
+namespace Clusters {
+
+namespace ThreadBorderRouterManagement {
+
+class ScopedThreadLock
+{
+public:
+    ScopedThreadLock() { DeviceLayer::ThreadStackMgr().LockThreadStack(); }
+    ~ScopedThreadLock() { DeviceLayer::ThreadStackMgr().UnlockThreadStack(); }
+};
+
+CHIP_ERROR GenericOpenThreadBorderRouterDelegate::Init()
+{
+    mCallback = nullptr;
+    ReturnErrorOnFailure(DeviceLayer::PlatformMgrImpl().AddEventHandler(OnPlatformEventHandler, reinterpret_cast<intptr_t>(this)));
+    // When the Thread Border Router is reboot during SetActiveDataset, we need to revert the active dateset.
+    ReturnErrorOnFailure(RevertActiveDataset());
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR GenericOpenThreadBorderRouterDelegate::GetBorderAgentId(MutableByteSpan & borderAgentIdSpan)
+{
+    otInstance * otInst = DeviceLayer::ThreadStackMgrImpl().OTInstance();
+    VerifyOrReturnError(otInst, CHIP_ERROR_INCORRECT_STATE);
```

Comment:
What error will this cause on the client?

Reply:
The error will be passed and returned by ServerInstance::Read(), and then the client will receive a failure report for this read action.

Reply:
Instead of returning CHIP_ERRORs that are assumed to encode to Failure, please encode IM-space errors everywhere.

Reply:
Let's follow-up

Reply:
#34595